docs(two-columns-layout-example): add products plugin

FabienMotte opened the pull request:
This PR adds the products plugin to the Autocomplete two columns layout example.

sarahdayan commented on examples/two-columns-layout/src/app.tsx:
```diff
@@ -13,21 +14,28 @@ autocomplete({
+      productsPlugin: products,
     } = elements;
 
+    const productsSection = products && (
+      <div className="aa-PanelSection--products">
+        <div className="aa-PanelSectionSource">{products}</div>
+      </div>
+    );
```
Couldn't we inline this instead of storing it in a variable?

FabienMotte replied:
I chose to store it in a variable because in the last PR, this section is displayed conditionally.
Should I inline it for now and refactor later?

sarahdayan replied:
Couldn't we have the condition inlined?

FabienMotte replied:
> Regarding the blurash placeholder, it will only show once Algolia has responded (not visible in your video). Until then we don't have the hash so we need to display something.

Is it possible that the state you mentioned is displayed at some point?
Items should not be visible until we get the results from Algolia, right?

Otherwise, we can simplify the states by just using a simple background color with a looping opacity animation when the image is not loaded and remove the Blurhash canvas, WDYT?

sarahdayan commented on examples/two-columns-layout/src/components/Blurhash.tsx:
```diff
@@ -0,0 +1,46 @@
+    const imageData = ctx.createImageData(width, height);
+    imageData.data.set(pixels);
+    ctx.putImageData(imageData, 0, 0);
+  });
```
This `useEffect` is missing dependencies.

sarahdayan commented on examples/two-columns-layout/src/components/Blurhash.tsx:
```diff
@@ -0,0 +1,46 @@
+  height = 128,
+  punch = 1,
+}: BlurhashProps) => {
+  const ref = useRef<HTMLCanvasElement>();
```
```suggestion
  const ref = useRef<HTMLCanvasElement>(null);
```

sarahdayan commented on examples/two-columns-layout/src/components/Blurhash.tsx:
```diff
@@ -0,0 +1,46 @@
+      ref={ref}
+      height={height}
+      width={width}
+      style={{
+        position: 'absolute',
+        top: 0,
+        bottom: 0,
+        left: 0,
+        right: 0,
+        width: '100%',
+        height: '100%',
+      }}
```
Could this be a CSS class?

sarahdayan commented on examples/two-columns-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,56 @@
+/** @jsx h */
+import { h } from 'preact';
+
+type IconProps = {
+  className?: string;
+};
```
```suggestion
```

sarahdayan commented on examples/two-columns-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,56 @@
+  className?: string;
+};
+
+export const ChevronRightIcon = ({ className }: IconProps) => (
```
Applicable to the other icons.

```suggestion
export const ChevronRightIcon = (props: React.SVGProps<SVGSVGElement>) => (
```

FabienMotte replied:
I went for `props: h.JSX.SVGAttributes<SVGSVGElement>`

sarahdayan commented on examples/two-columns-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,56 @@
+};
+
+export const ChevronRightIcon = ({ className }: IconProps) => (
+  <div className={['aa-ItemIcon aa-ItemIcon--noBorder', className].join(' ')}>
```
This should be outside of the icons.

sarahdayan commented on examples/two-columns-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,56 @@
+
+export const ChevronRightIcon = ({ className }: IconProps) => (
+  <div className={['aa-ItemIcon aa-ItemIcon--noBorder', className].join(' ')}>
+    <svg viewBox="0 0 24 24" fill="none" stroke="currentColor">
```
Applicable to the other icons.

```suggestion
    <svg viewBox="0 0 24 24" fill="none" stroke="currentColor" {...props}>
```

sarahdayan commented on examples/two-columns-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,56 @@
+  className?: string;
+};
+
+export const ChevronRightIcon = ({ className }: IconProps) => (
```
We use [function declarations](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function) in the codebase instead of expressions.

sarahdayan commented on examples/two-columns-layout/style.css:
```diff
@@ -60,9 +60,162 @@ body {
+  border-radius: 3px;
+  box-shadow: 0px 4px 8px rgba(35, 38, 59, 0.15);
+  cursor: pointer;
+  opacity: 0;
```
Why is it hidden on desktop?

Also, we should use the `hidden` HTML attribute instead since we don't use the `display` property.

FabienMotte replied:
On desktop, it is only visible when hovering an item. And on mobile, it is always visible (no hover state).
Should we change this behavior?

sarahdayan commented on examples/two-columns-layout/style.css:
```diff
@@ -60,9 +60,162 @@ body {
+  justify-content: space-between;
+}
+
+.aa-ProductItem .aa-ItemPictureBlurred {
+  position: absolute;
+  width: 100%;
+  height: 100%;
+  transition: opacity 0.2s ease-out, visibility 0s linear 0.2s;
+}
+
+.aa-ProductItem .aa-ItemPictureBlurred--loaded {
+  opacity: 0;
+  visibility: hidden;
+}
```
Shouldn't we have `.aa-ItemPicture--blurred and .aa-ItemPicture--loaded` instead?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  StarFillIcon,
+  FavoriteFillIcon,
+  FavoriteOutlineIcon,
+} from '../components/Icons';
```
We can probably have a barrel in `components/` and re-export everywhere, wdyt?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  },
+};
+
+const formatPrice = (val: number) => {
+  return val.toFixed(2).toLocaleString().replace(/\./g, ',');
+};
```
This mixes two notations, because `toLocaleString` uses the browser's localization to format the value, and the `replace` forces commas as decimal separators. On my French browser it works, but U.S. viewers will see some unorthodox notation.

Since the demos is in English and we usually go for American English everywhere, let's settle with U.S. notation, wdyt?

```suggestion
function formatPrice(value: number) {
  return value.toFixed(2).toLocaleString('en-US', { style: 'currency', currency: 'EUR' });
};
```

FabienMotte replied:
Thanks for the explanation!

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  hit: ProductHit;
+};
+
+const ProductItem = ({ hit }: ProductItemProps) => {
```
```suggestion
function ProductItem({ hit }: ProductItemProps) {
```

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  return (
+    <a href="#" className="aa-ItemLink aa-ProductItem">
```
If there's no `href`, should it be a link? If you change it to a `div` you can use `.aa-ItemWrapper` instead of `aa-ItemLink`.

FabienMotte replied:
It should be a link in a real use-case. I used a link to have the pointer cursor and add a hover transition but maybe I should just use a `div` indeed. WDYT?

sarahdayan replied:
Makes sense indeed if it's supposed to be a link. Maybe we could use `"https://example.org/"` which is designed for it?

FabienMotte replied:
I could use the link you mentioned. Should we `preventDefault` in `onClick` so we don’t leave the example by mistake?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+              loaded && 'aa-ItemPictureBlurred--loaded',
+            ]
+              .filter(Boolean)
+              .join(' ')}
```
We have a [`cx` utility](https://github.com/algolia/autocomplete/blob/467ea6f2a9903f11591368abeeb7a515ba150af0/examples/react-instantsearch-hooks/src/utils/cx.ts#L1) that we use in other examples, you could use it instead of rewriting this logic all the time.

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+                <div className="aa-ItemContentPriceDiscounted">
+                  {discountedPrice} {priceCurrency}
+                </div>
+              )}
```
With `toLocaleString`, no need for the currency (see my comment above). You can lighten your records 💪

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  const onLoad = useCallback(() => setLoaded(true), []);
+
+  const [favorite, setFavorite] = useState(false);
+  const onFavoriteClick = useCallback(
+    (event) => {
+      event.preventDefault();
+      event.stopPropagation();
+      setFavorite(!favorite);
+    },
+    [favorite]
+  );
```
Why `useCallback`?

FabienMotte replied:
It's not needed here, thanks!

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+
+const ProductItem = ({ hit }: ProductItemProps) => {
+  const [loaded, setLoaded] = useState(false);
+  const onLoad = useCallback(() => setLoaded(true), []);
```
Why `useCallback`?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+
+  const currentPrice = formatPrice(hit.price.value);
+  const discountedPrice = formatPrice(hit.price.discounted_value);
+  const priceCurrency = hit.price.currency === 'EUR' ? '€' : '$';
```
You won't need it (see below).

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  const discountedPrice = formatPrice(hit.price.discounted_value);
+  const priceCurrency = hit.price.currency === 'EUR' ? '€' : '$';
+
+  const rating = hit.reviews.rating;
+  const reviews = hit.reviews.count;
```
Do they need their own variables?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+          </div>
+        </div>
+
+        <button className="aa-ItemFavorite" onClick={onFavoriteClick}>
```
Couldn't the function be inlined?

sarahdayan commented on examples/two-columns-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,154 @@
+  const rating = hit.reviews.rating;
+  const reviews = hit.reviews.count;
+
+  const stars = [];
+  for (let i = 0; i < 5; i++) {
+    const Star = i >= rating ? StarOutlineIcon : StarFillIcon;
+    stars.push(<li key={i}>{<Star className="aa-StarIcon" />}</li>);
+  }
```
This logic could go in the return function in a `stars.map`.

sarahdayan reviewed (commented):
I've noticed a few things in the UI:

- The title "Products" is very discreet. Why not use the [header template](https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-theme-classic/#header) from the theme?
- When there are less than 3 products, they fill the space, meaning they grow in height too. Couldn't we instead keep them at the same size?
- When images are loading, the space isn't "reserved" so there's layout shift. Also, with the blurhash we likely have three possible states: empty image, blurred image and crisp image instead of going from not loaded to loaded. Do we think it's good UX?
- When resizing the browser (large to narrow) there's a moment when the panel has a weird width (while resizing). It's particularly visible on Firefox. Is this caused by the custom CSS, or was it already there?

sarahdayan commented:
Another UI comment: I'm not a big fan of the color and styles for the star rating system, I find the blue a bit too "system" and it overpowers the rest.

Here's a suggestion, wdyt? These are the colors we use for the star rating system on the docs' solutions.

![Capture d’écran 2022-03-08 à 11 01 28](https://user-images.githubusercontent.com/5370675/157213923-1a99bb80-1810-487e-813f-c5bc9f9ec21a.png)

FabienMotte commented:
Thanks for your comments! 🙏 

> I've noticed a few things in the UI:
> 
> * The title "Products" is very discreet. Why not use the [header template](https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-theme-classic/#header) from the theme?

It should be better like this. I removed the line at the end because the title is using a white background, conflicting with the background color on all products added by the smart preview feature.
I'll see how to handle this specific case in the relevant PR.

<img width="890" alt="image" src="https://user-images.githubusercontent.com/662153/157309393-96d275ea-db50-4414-855d-52559129c3a4.png">

> * When there are less than 3 products, they fill the space, meaning they grow in height too. Couldn't we instead keep them at the same size?

Indeed, thanks!

<img width="888" alt="image" src="https://user-images.githubusercontent.com/662153/157309917-aabdbd55-7aad-4cea-b3c6-3b54ab12fbb9.png">

> * When images are loading, the space isn't "reserved" so there's layout shift. Also, with the blurhash we likely have three possible states: empty image, blurred image and crisp image instead of going from not loaded to loaded. Do we think it's good UX?

When the image wasn't loaded, the body part was taking all the available height, hiding the Blurhash canvas. Now it should work properly. The blur placeholder is not an image, so it should be visible as the same time as the results.
I added a timeout + network throttling to see the states in the below video.

https://user-images.githubusercontent.com/662153/157309965-46c01d13-f23d-408a-8c03-7603429c891c.mov

> * When resizing the browser (large to narrow) there's a moment when the panel has a weird width (while resizing). It's particularly visible on Firefox. Is this caused by the custom CSS, or was it already there?

I think it is due to media queries that update the `max-width` of the container, I removed them, can you try again?

FabienMotte commented:
> Another UI comment: I'm not a big fan of the color and styles for the star rating system, I find the blue a bit too "system" and it overpowers the rest.
> 
> Here's a suggestion, wdyt? These are the colors we use for the star rating system on the docs' solutions. I also reduced the descriptive text and used a more muted color.
> 
> ![Capture d’écran 2022-03-08 à 11 05 17](https://user-images.githubusercontent.com/5370675/157214423-e74db64c-c8e0-43bc-a184-debfdb7b432a.png)

It's looking much better, thank you (and using one less SVG)!

<img width="211" alt="image" src="https://user-images.githubusercontent.com/662153/157311219-94c3bf52-ac29-42d8-88d7-9c78c1b9d385.png">

sarahdayan commented on examples/two-column-layout/src/components/Icons.tsx:
```diff
@@ -0,0 +1,39 @@
+  );
+}
+
+export function FavoriteFillIcon(props: h.JSX.SVGAttributes<SVGSVGElement>) {
+  return (
+    <svg viewBox="0 0 24 24" fill="currentColor" {...props}>
+      <path d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z" />
+    </svg>
+  );
+}
+
+export function FavoriteOutlineIcon(props: h.JSX.SVGAttributes<SVGSVGElement>) {
+  return (
+    <svg viewBox="0 0 24 24" fill="currentColor" {...props}>
+      <path d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z" />
+    </svg>
+  );
+}
```
Could this be a single SVG? I see the `path` is different but I guess we could have the same one, either using `fill="none" stroke="currentColor"` or the opposite. wdyt?

sarahdayan commented on examples/two-column-layout/style.css:
```diff
@@ -19,10 +19,12 @@ body {
-  color: rgb(var(--aa-muted-color-rgb));
-  font-weight: normal;
+/* Blurhash */
+.BlurhashCanvas {
```
```suggestion
.aa-BlurhashCanvas {
```

sarahdayan commented on examples/two-column-layout/style.css:
```diff
@@ -59,9 +61,173 @@ body {
+  height: auto;
+  border-radius: 3px;
+  transition: transform 1.8s ease-out;
+  transform-origin: center bottom;
```
I find the zoom-in transition from the bottom a bit weird, but it might just be me.

FabienMotte replied:
I did this because some pictures are aligned at the bottom, like shoes for example, but I guess it could work with `center` too

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+    hit.price.currency
+  );
+
+  const stars = Array(5)
+    .fill(null)
+    .map((_, i) => {
+      return (
+        <li key={i}>
+          <div
+            className={cx(
+              'aa-ItemIcon aa-ItemIcon--noBorder aa-StarIcon',
+              i >= hit.reviews.rating && 'aa-StarIcon--muted'
+            )}
+          >
+            <StarIcon />
+          </div>
+        </li>
+      );
+    });
```
We don't reuse `stars`, so let's inline it on the return statement.

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+
+  const stars = Array(5)
+    .fill(null)
+    .map((_, i) => {
```
```suggestion
    .map((_, index) => {
```

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+    .fill(null)
+    .map((_, i) => {
+      return (
+        <li key={i}>
```
```suggestion
        <li key={index}>
```

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+          <div
+            className={cx(
+              'aa-ItemIcon aa-ItemIcon--noBorder aa-StarIcon',
+              i >= hit.reviews.rating && 'aa-StarIcon--muted'
```
```suggestion
              index >= hit.reviews.rating && 'aa-StarIcon--muted'
```

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+  const FavoriteIcon = favorite ? FavoriteFillIcon : FavoriteOutlineIcon;
+
+  const currentPrice = formatPrice(hit.price.value, hit.price.currency);
+  const discountedPrice = formatPrice(
+    hit.price.discounted_value,
+    hit.price.currency
+  );
```
We don't reuse `discountedPrice`, so let's inline it on the return statement.

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+  const [favorite, setFavorite] = useState(false);
+  const FavoriteIcon = favorite ? FavoriteFillIcon : FavoriteOutlineIcon;
+
+  const currentPrice = formatPrice(hit.price.value, hit.price.currency);
```
We don't reuse `currentPrice`, so let's inline it on the return statement.

sarahdayan commented on examples/two-column-layout/src/plugins/productsPlugin.tsx:
```diff
@@ -0,0 +1,168 @@
+          onClick={(event) => {
+            event.preventDefault();
+            event.stopPropagation();
+            setFavorite(!favorite);
```
```suggestion

            setFavorite((currentFavorite) => !currentFavorite);
```

seafoox commented:
- Can we add a "See all products (x)" as done in the "[Desktop] Multi category results" view in the Ecom UI Design Kit?
- Same for the "See all" displayed next to the articles

- Could we add some debouncing so it reduces this flickering effect, while keeping it on the left items composed of the QuerySuggestions. So this way the end users can better focus on the left part, before focusing on the right part later on.

- Could we work on adding the highlighting on the product hits?
